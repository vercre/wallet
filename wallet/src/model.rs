//! Model for the wallet application state.

pub mod credential;
mod issuance;

pub use credential::CredentialState;
use issuance::IssuanceState;
use serde::Serialize;

use super::Aspect;
use crate::capabilities::store::{Store, StoreEntry};

/// State for the wallet application.
#[derive(Default, Serialize)]
pub struct Model {
    /// Which aspect of the application is currently active.
    pub active_view: Aspect,

    /// Credential state.
    pub credential: CredentialState,

    /// Issuance state.
    pub issuance: Option<IssuanceState>,

    /// Error state.
    pub error: Option<String>,
}

/// Methods to set the application state based on events.
impl Model {
    /// An error has occurred. Set the error state.
    pub fn error(&mut self, error: String) {
        self.active_view = Aspect::Error;
        self.error = Some(error);
    }

    /// Set up the model with an initial state.
    pub fn ready(&mut self) {
        self.active_view = Aspect::CredentialList;
        self.credential = CredentialState::init();
    }

    /// The user has selected a credential in their wallet to view.
    pub fn select_credential(&mut self, id: String) {
        self.active_view = Aspect::CredentialDetail;
        self.credential.id = Some(id);
    }

    /// The credentials have been retrieved from the wallet's store.
    pub fn credentials_loaded(&mut self, entries: Vec<StoreEntry>) {
        self.active_view = Aspect::CredentialList;
        self.credential.set_credentials(entries);
    }

    /// The user has deleted a credential from their wallet.
    pub fn delete_credential(&mut self) {
        self.credential.id = None;
    }

    /// The user wants to scan an issuance offer QR code.
    pub fn scan_issuance_offer(&mut self) {
        self.active_view = Aspect::IssuanceScan;
        self.issuance = None;
    }

    /// The user has scanned an issuance offer QR code so we can initiate a
    /// pre-authorized issuance flow.
    pub fn issuance_offer<Ev>(
        &mut self, encoded_offer: &str, http: crux_http::Http<Ev>, store: Store<Ev>,
    ) {
        self.active_view = Aspect::IssuanceOffer;
        match IssuanceState::from_offer(encoded_offer, http, store) {
            Ok(issuance_state) => self.issuance = Some(issuance_state),
            Err(e) => self.error(e.to_string()),
        }
    }
}
