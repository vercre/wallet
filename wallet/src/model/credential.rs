//! Credential sub-app state.

use serde::{Deserialize, Serialize};
use vercre_holder::credential::Credential;

use crate::capabilities::store::StoreEntry;

/// Application state for the credential sub-app.
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[allow(clippy::module_name_repetitions)]
pub struct CredentialState {
    /// Currently selected credential.
    pub id: Option<String>,

    /// Credentials stored in the wallet.
    pub credentials: Vec<Credential>,
}

impl CredentialState {
    /// Create a new credential state.
    #[must_use]
    pub fn init() -> Self {
        Self {
            id: None,
            credentials: vec![],
        }
    }

    /// Set the credential list from a set of StoreEntries.
    pub fn set_credentials(&mut self, entries: Vec<StoreEntry>) {
        let mut credentials = vec![];
        for entry in entries {
            if let StoreEntry::Data(bytes) = entry {
                let credential: Credential =
                    serde_json::from_slice(&bytes).expect("should deserialize");
                credentials.push(credential);
            }
        }
        self.credentials = credentials;
    }
}
