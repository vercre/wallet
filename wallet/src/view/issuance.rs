//! Issuance flow view models.

use serde::{Deserialize, Serialize};
use vercre_holder::CredentialConfiguration;

use crate::model::IssuanceState;

/// View-friendly representation of a credential configuration.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct CredentialSummary {
    /// Credential configuration identifier.
    pub config_id: String,

    /// Name from issuer's metadata.
    pub name: String,

    /// Description from issuer's metadata.
    pub description: String,

    /// Claims the credential contains. The entries are the claim names. For
    /// nested claims, the claim name is prefixed with the parent claim name.
    pub claims: Vec<String>,

    /// Background color from issuer's metadata. Empty string if not applicable.
    pub background_color: String,

    /// Text color from issuer's metadata. Empty string if not applicable.
    pub text_color: String,

    /// URL for a logo image.
    ///
    /// Empty string if not applicable.
    pub logo_url: String,

    /// URL for a background image.
    ///
    /// Empty string if not applicable.
    pub background_url: String,
}

impl CredentialSummary {
    /// Convert a CredentialConfiguration to a CredentialSummary.
    pub fn from_configuration(
        config_id: impl Into<String>, config: &CredentialConfiguration,
    ) -> Self {
        let (name, description, background_color, text_color, logo_url, background_url) =
            match &config.display {
                Some(display_list) => {
                    // TODO: Locale support.
                    let display = display_list
                        .clone()
                        .into_iter()
                        .find(|display| display.locale.is_none())
                        .unwrap_or_else(|| display_list[0].clone());
                    let logo_url = match display.logo {
                        Some(logo) => logo.uri.unwrap_or_default(),
                        None => String::new(),
                    };
                    let background_url = match display.background_image {
                        Some(background) => background.uri.unwrap_or_default(),
                        None => String::new(),
                    };
                    (
                        display.name,
                        display.description.unwrap_or_default(),
                        display.background_color.unwrap_or_default(),
                        display.text_color.unwrap_or_default(),
                        logo_url,
                        background_url,
                    )
                }
                None => (
                    String::new(),
                    String::new(),
                    String::new(),
                    String::new(),
                    String::new(),
                    String::new(),
                ),
            };
        Self {
            config_id: config_id.into(),
            name,
            description,
            claims: config.claims_display(None), // TODO: Locale from shell.
            background_color,
            text_color,
            logo_url,
            background_url,
        }
    }
}

/// View-friendly representation of a transaction code specification.
#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
pub struct TxCode {
    /// The type of characters expected. Will be "numeric" or "text".
    pub input_mode: String,

    /// The number of characters expected. Zero if not applicable.
    pub length: i32,

    /// Helper text to display to the user.
    pub description: String,
}

impl Default for TxCode {
    fn default() -> Self {
        Self {
            input_mode: "numeric".into(),
            length: 0,
            description: "".into(),
        }
    }
}

impl From<Option<vercre_holder::TxCode>> for TxCode {
    fn from(tx_code: Option<vercre_holder::TxCode>) -> Self {
        match tx_code {
            Some(tx_code) => Self {
                input_mode: tx_code.input_mode.unwrap_or("numeric".into()),
                length: tx_code.length.unwrap_or_default(),
                description: tx_code.description.unwrap_or_default(),
            },
            None => Self::default(),
        }
    }
}

/// View model for an issuance flow.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct IssuanceView {
    /// The credential issuer's identifier.
    pub issuer: String,

    /// The credential issuer's name.
    pub issuer_name: String,

    /// Credential being offered.
    pub offered: CredentialSummary,

    /// PIN requirements.
    pub tx_code: TxCode,
}

impl From<Option<IssuanceState>> for IssuanceView {
    fn from(state: Option<IssuanceState>) -> Self {
        match state {
            Some(state) => {
                // Pick the first credential in the issuance state to display. Since the
                // app is intended to allow the holder to accept or reject the
                // credential and/or claims one at a time.
                if state.offered.is_empty() {
                    return Self::default();
                }
                let Some(next_config) = state.offered.iter().next() else {
                    return Self::default();
                };
                Self {
                    issuer: state.issuer.clone(),
                    issuer_name: state.issuer_name.clone(),
                    tx_code: state.tx_code.into(),
                    offered: CredentialSummary::from_configuration(next_config.0, next_config.1),
                }
            }
            None => Self::default(),
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot;

    use super::*;

    #[test]
    fn credential_config() {
        let json = serde_json::json!({
            "format": "jwt_vc_json",
            "scope": "EmployeeIDCredential",
            "cryptographic_binding_methods_supported": [
                "did:key",
                "did:web"
            ],
            "credential_signing_alg_values_supported": [
                "ES256K",
                "EdDSA"
            ],
            "proof_types_supported": {
                "jwt": {
                    "proof_signing_alg_values_supported": [
                        "ES256K",
                        "EdDSA"
                    ]
                }
            },
            "display": [
                {
                    "name": "Employee ID",
                    "description": "Vercre employee ID credential",
                    "locale": "en-NZ",
                    "logo": {
                        "uri": "https://vercre.github.io/assets/employee.png",
                        "alt_text": "Vercre Logo"
                    },
                    "text_color": "#ffffff",
                    "background_color": "#323ed2",
                    "background_image": {
                        "uri": "https://vercre.github.io/assets/employee-background.png",
                        "alt_text": "Vercre Background"
                    }
                }
            ],
            "credential_definition": {
                "type": [
                    "VerifiableCredential",
                    "EmployeeIDCredential"
                ],
                "credentialSubject": {
                    "email": {
                        "mandatory": true,
                        "value_type": "string",
                        "display": [
                            {
                                "name": "Email",
                                "locale": "en-NZ"
                            }
                        ]
                    },
                    "family_name": {
                        "mandatory": true,
                        "value_type": "string",
                        "display": [
                            {
                                "name": "Family name",
                                "locale": "en-NZ"
                            }
                        ]
                    },
                    "given_name": {
                        "mandatory": true,
                        "value_type": "string",
                        "display": [
                            {
                                "name": "Given name",
                                "locale": "en-NZ"
                            }
                        ]
                    },
                    "address": {
                        "street_address": {
                            "value_type": "string",
                            "display": [
                                {
                                    "name": "Street Address",
                                    "locale": "en-NZ"
                                }
                            ]
                        },
                        "locality": {
                            "value_type": "string",
                            "display": [
                                {
                                    "name": "Locality",
                                    "locale": "en-NZ"
                                }
                            ]
                        },
                        "region": {
                            "value_type": "string",
                            "display": [
                                {
                                    "name": "Region",
                                    "locale": "en-NZ"
                                }
                            ]
                        },
                        "country": {
                            "value_type": "string",
                            "display": [
                                {
                                    "name": "Country",
                                    "locale": "en-NZ"
                                }
                            ]
                        }
                    }
                }
            }
        });

        let config: CredentialConfiguration =
            serde_json::from_value(json.clone()).expect("should deserialize from json");
        let summary = CredentialSummary::from_configuration("test", &config);
        assert_yaml_snapshot!("config_summary", summary, {
            ".claims" => insta::sorted_redaction(),
        });
    }
}
