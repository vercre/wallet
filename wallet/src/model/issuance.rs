//! Issuance sub-app state.
use std::collections::HashMap;

use anyhow::bail;
use vercre_holder::credential::ImageData;
use vercre_holder::issuance::{
    Accepted, IssuanceFlow, NotAccepted, PreAuthorized, WithOffer, WithToken, WithoutToken,
};
use vercre_holder::{CredentialConfiguration, CredentialOffer, PreAuthorizedCodeGrant};

use crate::provider::Provider;

/// Configuration and image information for an offered credential.
#[derive(Clone, Debug, Default)]
pub struct OfferedCredential {
    /// Credential configuration.
    pub config: CredentialConfiguration,

    /// Logo image data.
    pub logo: Option<ImageData>,

    /// Background image data.
    pub background: Option<ImageData>,

    /// Credential has been received.
    pub received: bool,

    /// Credential has been stored.
    pub stored: bool,
}

impl OfferedCredential {
    /// Determine if the credential logo needs to be fetched.
    pub fn needs_logo(&self) -> bool {
        if self.logo.is_some() {
            return false;
        }
        if let Some(display) = &self.config.display {
            if let Some(logo) = &display[0].logo {
                if logo.uri.is_some() {
                    return true;
                }
            }
        }
        false
    }

    /// Determine if the credential background needs to be fetched.
    pub fn needs_background(&self) -> bool {
        if self.background.is_some() {
            return false;
        }
        if let Some(display) = &self.config.display {
            if let Some(background) = &display[0].background_image {
                if background.uri.is_some() {
                    return true;
                }
            }
        }
        false
    }
}

/// Application state for the issuance sub-app.
#[derive(Clone, Debug, Default)]
#[allow(clippy::module_name_repetitions)]
#[allow(clippy::large_enum_variant)]
pub enum IssuanceState {
    /// No issuance is in progress.
    #[default]
    Inactive,

    /// An offer has been received
    Offered { offer: CredentialOffer, grant: PreAuthorizedCodeGrant },

    /// Issuer metadata has been received. Can use this state to keep updating
    /// the offered credentials' logo and background images.
    IssuerMetadata {
        flow: IssuanceFlow<WithOffer, PreAuthorized, NotAccepted, WithoutToken>,
        offerred: HashMap<String, OfferedCredential>,
    },

    /// The offer has been accepted by the user. Can use this state to update
    /// the PIN number if needed.
    Accepted {
        flow: IssuanceFlow<WithOffer, PreAuthorized, Accepted, WithoutToken>,
        offerred: HashMap<String, OfferedCredential>,
    },

    /// An access token has been received.
    Token {
        flow: IssuanceFlow<WithOffer, PreAuthorized, Accepted, WithToken>,
        offerred: HashMap<String, OfferedCredential>,
    },

    /// A proof has been created. Can use this state to receive credentials and
    /// update the offered list to keep track of outstanding credentials. Can
    /// also use it to keep track of the credentials stored.
    Proof {
        flow: IssuanceFlow<WithOffer, PreAuthorized, Accepted, WithToken>,
        offerred: HashMap<String, OfferedCredential>,
        proof: String,
    },
}

/// State change implementation.
impl IssuanceState {
    /// Create an issuance state from a URL-encoded offer.
    pub fn from_offer(encoded_offer: &str) -> anyhow::Result<Self> {
        let offer_str = urlencoding::decode(encoded_offer)?;
        let offer = serde_json::from_str::<CredentialOffer>(&offer_str)?;

        // Check the offer has a pre-authorized grant. This is the only flow
        // type supported by this wallet (for now).
        let Some(pre_auth_code_grant) = offer.pre_authorized_code() else {
            bail!("grant other than pre-authorized code is not supported");
        };

        Ok(Self::Offered {
            offer,
            grant: pre_auth_code_grant,
        })
    }

    /// Cancel the issuance process.
    pub fn cancel<Ev>(&mut self, _provider: &Provider<Ev>) -> anyhow::Result<()>
    where
        Ev: 'static,
    {
        // TODO: Reset state
        Ok(())
    }
}
