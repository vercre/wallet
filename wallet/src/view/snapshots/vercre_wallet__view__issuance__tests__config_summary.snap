---
source: wallet/src/view/issuance.rs
assertion_line: 296
expression: summary
snapshot_kind: text
---
config_id: test
name: Employee ID
description: Vercre employee ID credential
claims:
  - Address.Country
  - Address.Locality
  - Address.Region
  - Address.Street Address
  - Email
  - Family name
  - Given name
background_color: "#323ed2"
text_color: "#ffffff"
logo_url: "https://vercre.github.io/assets/employee.png"
background_url: "https://vercre.github.io/assets/employee-background.png"
